Shade external dependencies in the rest client jar

hub-cap opened the pull request:
This commit removes all external dependencies from the rest client jar,
and shades them in an 'internal.' package within the jar, using
shadowJar gradle plugin. All projects that depended on the existing jar
have been converted to using the 'internal.' package prefixes to
interact with the rest client.

Closes #25208

hub-cap commented on client/rest/build.gradle:
```diff
@@ -38,13 +42,50 @@ publishing {
+  }
+}
+
+// Actual task for
```
heh, will fix this comment

javanna commented on client/test/build.gradle:
```diff
@@ -27,7 +26,7 @@ targetCompatibility = JavaVersion.VERSION_1_7
 
 dependencies {
-  compile "org.apache.httpcomponents:httpcore:${versions.httpcore}"
+  compile "org.elasticsearch.client:elasticsearch-rest-client:${version}"
```
why is httpcore replaced with the low-level rest client here?

javanna replied:
I think I get it. either we use the ordinary httpcore classes, or the internal ones but they come from our own big jar.

hub-cap replied:
well, we really cant use ordinary ones now. Since all of the dependencies in rest client are shaded to use internal.org.apache classes, if we leak anything like `Header` into an API, we need to use the internal.org.apache since those are in the method signatures. So, we have to use our internal ones always, which is why we removed the httpcore stuff, because client/test now uses `internal.o.a.h.Header`, for example.

javanna replied:
thanks for the explanation!

javanna commented on core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy:
```diff
@@ -69,6 +69,16 @@ grant codeBase "${codebase.elasticsearch-rest-client-6.0.0-beta1-SNAPSHOT.jar}"
+  // rest makes socket connections for rest tests
+  permission java.net.SocketPermission "*", "connect";
+};
+
```
don't we need to remove the existing jar entry without the suffix?

hub-cap replied:
Hopefully this will be explained better in the commit msg update

rjernst replied:
I also do not understand why both of these are necessary. The rest client used by the test framework for REST integ tests should be using the jar with no classifier. The deps version above should replace the 2 http grants below? I don't see where the nodeps version would come in.

hub-cap replied:
It doesn't, anymore :)

hub-cap replied:
I lied, it does, in an IDE /sadface

nik9000 replied:
So we need the grant above for gradle and the grant below for the IDEs which don't use the fully shaded dependency. They use the -nodeps jar. I *think* they should be using the shadedDeps jar rather than the original dependencies. That way the packages line up.

hub-cap replied:
Id love for them to use the shaded jar itself, but we use project dependencies, and intellij is smarter than me when it comes to dealing w/ dependencies. here is what happens to the other projects, and they seem to (only in intellij) depend on the jar task w/ the funky classifier, even though I remove it from the runtime artifacts and remove the extendsFrom compile. So these stay, just for intellij purposes. And they are using the same packages anyway, since the -deps jar and the no classifier actual shaded jar have the same `relocate` bits.

![image](https://user-images.githubusercontent.com/613352/28525165-ec1b6064-7048-11e7-9bf5-ff4d0b33141c.png)

hub-cap commented:
@javanna Ill leave some better comments in the github commit on why the artifacts are split out like that. its somewhat confusing.

hub-cap commented:
ok I decided to just add better docs to the build.gradle instead. Its saner there anyway.

rjernst commented on client/rest/build.gradle:
```diff
@@ -38,13 +55,50 @@ publishing {
+shadowJar {
+  configurations = [project.configurations.shade]
+  classifier = null
+  relocate 'org.apache', 'internal.org.apache'
```
Shouldn't there be no need to use `relocate` here? If there is then unit tests ran with unshaded classes, which is exactly what we were trying to avoid.

hub-cap replied:
This is the duplicated shadowing task that actually shades the classes in src as well as the `internal.` things.

rjernst replied:
That is my point. We shade the dependency itself. Nothing in src should be referring to unshaded packages, so there should be no need to do t his relocation here.

javanna commented on client/rest/build.gradle:
```diff
@@ -19,6 +19,19 @@
+ *    needs compilation and depends on `internal.` classes, so an intermediary jar needs to exist to satisfy the compile.
+ *    The `deps` classifier is added to this jar.
+ * 3) The *actual* jar that will be used by clients. This has no classifier, contains the rest client src and `internal.`.
+ */
```
thanks for adding this comment!

hub-cap commented:
@javanna @rjernst I have cleaned up the build so that it uses the actual shaded jar outside of the rest project, instead of relying on the transient project dependencies (which was the -deps.jar). I have also removed the -nodeps jar from existence outside of the build, so the shaded jar is the only thing that other projects know about / build to.

rjernst commented:
Just a note about this in general: I think as a followup (for rc/GA) we need to pull javadocs from the deps and add them to the generated client jar javadocs.

nik9000 commented:
> also think we should think of a better package name prefix, since these are part of the api

We could, in the grand tradition of Java, go super verbose with something like `org.elasticsearch.client.shaded.httpcomponents`.

nik9000 commented on client/rest/build.gradle:
```diff
@@ -19,6 +19,29 @@
 
+/**
+ * The rest client is a shaded jar. It contains the source of the rest client, as well as all the dependencies,
+ * shaded to the `internal.` package. 3 artifacts come out of this build process.
```
`internal.` is no longer accurate.

nik9000 commented on client/rest/build.gradle:
```diff
@@ -19,6 +19,29 @@
+ * 3) The *actual* jar that will be used by clients. This has no classifier, contains the rest client src and `internal.`.
+ *    This jar is the only actual output artifact of this job.
+ *
+ *  The IDE does not like removing artifacts and changing configurations on the fly, so the bits that make the build
```
s/The IDE/IDEs/ ? I expect you mean IntelliJ and Eclipse.

nik9000 commented on client/rest/build.gradle:
```diff
@@ -19,6 +19,29 @@
+ *    The reason this jar was not overwritten was because gradle knows what its inputs and outputs are for a given task
+ *    and we should not alter that by overwriting the original jar. What we do is add a `nodeps` classifier to it instead.
+ *    This jar is removed from the output artifacts in cli, and retained in IDE builds.
+ * 2) A jar that contains *only* the `internal.` dependencies. This is a jar that is built before the src is compiled.
```
I like this comment but I'd rephrase it, maybe moving the number 2 jar to the number 1 jar and explaining that it is also there for IDEs. I think it'd be good also to explain that we use the shaded package names in the source so we still jar up the bits that come out of javac without any post processing.

nik9000 commented on build.gradle:
```diff
@@ -268,7 +268,7 @@ subprojects {
   allprojects {
-    if (project.path == ':test:framework') {
+    if (project.path == ':test:framework' || project.path == ':client:test') {
       // :test:framework:test cannot run before and after :core:test
```
Can you add a comment explaining why `:core:test` also cannot run before `:core:test`?

hub-cap replied:
Ill leave that for ryan to answer heh

rjernst replied:
I don't understand your statement @nik9000. The task graph cannot have cycles, and a dependency on yourself is still a cycle.

nik9000 replied:
Sorry, `:client:test` cannot run before `:core:test`. Basically, if we are going to leave a comment explaining why `:test:framework` needs this exemption we should probably explain the `:client:test` too.

rjernst replied:
Ok I see the confusion. The comment should be updated. It means that if the code below this were to run, a circular dependency would be added between `:test:framework:test` and `:core:test`. Likewise, the new check is added because `:client:test` would have a circular task dep with `:client:rest:test`.

nik9000 commented on client/rest/build.gradle:
```diff
@@ -38,13 +61,66 @@ publishing {
+shadowJar {
+  configurations = [project.configurations.shade]
+  classifier = null
+  relocate 'org.apache', 'org.elasticsearch.client'
```
I think it'd be clearer to do this with the regular jar task because you are just combining the two jar files into one, right? At least, I thought that is what we'd talked about doing.

hub-cap replied:
You are right, at this point I could combine the deps jar with the src/ in the project. I can follow up with this, as its a better idea than running another "very similar" shadow task. I can follow up with this

nik9000 commented on client/rest/build.gradle:
```diff
@@ -38,13 +61,66 @@ publishing {
+}
+
+jar {
+  // rename the default jar because we are removing the classifier from the shadowJar, which is a null classifier
```
I'd say something like "move the default jar to another classifier so the jar with the dependencies can have the empty classifier".

hub-cap commented on client/rest/build.gradle:
```diff
@@ -56,6 +132,8 @@ dependencies {
   signature "org.codehaus.mojo.signature:java17:1.0@signature"
 }
 
+dependencyLicenses.dependencies = project.configurations.shade
```
Ill chat w/ ryan about this, good call for bringing it up

rjernst replied:
Good idea

rjernst commented:
> We could, in the grand tradition of Java, go super verbose with something like org.elasticsearch.client.shaded.httpcomponents

Please let's not do that. The package names have real costs (eg Windows file path limits).

nik9000 commented:
> Please let's not do that. The package names have real costs (eg Windows file path limits).

Fine by me. I'd certainly prefer it if all the shaded things were grouped together in some obvious way but I'm not going to block anything over it.

hub-cap commented:
👍  ill create a few issues off this initial one based on the feedback here

javanna commented:
thanks @hub-cap ! Was this supposed to go to 5.x too? I see it has the 5.6 label.

hub-cap commented:
yes i have backported it, ty for keepin an eye out @javanna <3
